### PR TITLE
Improve the autocomplete feature of the builtin python IDE

### PR DESCRIPTION
**Pull request by dimven**
### Purpose

I've been having some instabilities when using the built in IDE with large scripts and for extended periods of time and I retraced that back to the current autocomplete implementation. I've revised the current implementation so that it avoids raising as many exceptions as it did. I then tried to improve the type guessing as much as possible.

The type guessing should be relatively accurate for simple statements. We can never guarantee a 100% coverage for complicated variable definitions, because we don't have the luxury of running in the same scope as the actual script or pre-executing the entire code of the script.

Here's a before capture; notice the constant exceptions being thrown in the console:
current autocomplete

And this is with the proposed changes:
new autocomplete

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner 

### FYIs

This entire class is pretty independent of the rest of Dynamo, so it should work fine in both current and future versions of Dynamo.

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@dimven can you try to improve the diff here? It looks like there are a bunch of white space only changes, or is that much of the code actually new?

**Comment by dimven** [comment]
👍 Not sure why this happened and why git is detecting white spaces as a change. I must have "Ctrl+I"-ed the whole text at some point. I'll fix it over the weekend and resubmit

**Comment by dimven** [comment]
Now that I'm back from the holidays, I did some more testing and still see the slowdowns and instability whenever I use a `from XYZ import ABC` type of import statements. I think I'll need to reevaluate those as well

**Comment by Dewb** [comment]
Hi @dimven, thanks for tackling this. Any incremental improvements you can make to the autocompletion engine would be great! It also might be worth looking into replacing the entire internal engine with a complete Python autocompletion solution like Jedi -- it could in the end be less work than wrestling with regexes.

Both Atom and VS Code use Jedi for their Python autocompletion. Here are the cores of their implementations:
https://github.com/brennv/autocomplete-python-jedi/blob/master/lib/provider.coffee
https://github.com/DonJayamanne/pythonVSCode/blob/master/src/client/providers/completionSource.ts

Atom's provider API is in Coffeescript, and VS Code's is in Typescript, while Jedi is implemented in Python itself. Both of these extensions just launch a Python process to run Jedi and communicate with it from the editor's language. I don't see any reason why the same couldn't be done from C#.

Of course, this strategy is not without its pitfalls. It will probably be easiest to run Jedi in normal Python, and then you could run into challenges due to autocomplete running in normal Python while Dynamo actually runs IronPython. (In that case, this might help: https://github.com/gtalarico/ironpython-stubs)

So this could be a much bigger job than just tweaking regexes, but I thought I'd mention it! 

Another worthwhile line of investigation might be adding the ability to select an external editor for Python nodes, so editing them can pop up a window in your favorite IDE, which presumably already has Python autocompletion (though the normal Python vs. IronPython issues might still apply.)

**Comment by jnealb** [comment]
@Dewb @dimven @mjkkirschner I vote the very last suggestion :+1:

**Comment by radumg** [comment]
I also vote for the last suggestions, had that discussion in another issue too : https://github.com/DynamoDS/Dynamo/issues/6513#issuecomment-280946142

**Comment by dimven** [comment]
My view on this aligns with @andydandy74 's - unless Dynamo is properly set up and distributed with a default external IDE out of the box (i.e. how  #develop environment used in Revit macros), it would be unpractical to completely forgo the current approach. Otherwise we're just asking for more deployment hell.

Without a doubt, it would be great to have an easy to customize access to any IDE but that shouldn't compromise the experience of the majority of the users.

On a side note, why was #6513 closed ?

**Comment by dimven** [comment]
- Docstrings are automatically excluded from the code analysis. This is a really big improvement for cases where we handle large multiline string blocks in our code.
- The autocomplete scope now picks up references to external assemblies and can support additional imports.
- The default location of the python builtin library is added to the module search path and now those will be picked up too
- Some of the regex syntax has been revised and instantiated as compiled to hopefully improve performance. 

TODO:
- I'd like to further rework the regex syntax for variable detection and the import statements
- Add support for basic variable unpacking(i.e. a, b, c = 1, 2, 3)

**Comment by dimven** [comment]
@mjkkirschner just pushed a big update and would love to hear what you think about it:

- All completion tests are now outdated and need to be revised!
- All regex statements have been revised for speed and responsiveness
- Import statement detection is much more robust and powerful and can have all kinds of custom names
- We now have variable assignment unpacking



The autocomplete now feels snappier and more responsive with larger scripts too:

**Comment by Racel** [comment]
@dimven - This is awesome. We will review and get back to you. But just so you know, we are currently heads down on Dynamo 2.0, and this may have to wait for a 2.1 release. Thanks for your patience. We will get back to you soon.

**Comment by dimven** [comment]
@Racel no hurry on my end :) . I'm pretty happy with how the new code behaves on my end and the feedback I got from others is positive, so I don't think I'll modify it any further for the time being.

I also got around to implementing a very rudimentary overall autocomplete functionality:





but I'll first have to make sure it works well with the changes in 2.0 and will start a separate PR for it.

**Comment by phliberato** [comment]
@Racel do you have any forecast about when it will be released Dynamo 2.0?

**Comment by Dewb** [comment]
Thanks @dimven! I've skimmed your changes and they look pretty good. If the current implementation is causing crashes like #8931 I think that raises the urgency on this. I'll talk to folks and see what we need to do to move this forward.

**Comment by andydandy74** [comment]
Would be great to have this in the next stable release...

**Comment by mjkkirschner** [comment]
I will find time to look at this more closely as soon as I can. I'm sorry it has been open so long without a solid review.

 One thing I see immediately is the altering of public methods which we can't do without evaluating the risk - as it will break binary compatibility - we can keep these old public methods, mark them obsolete and create new ones though - and where possible we should be more conservative with access modifiers.

Another initial thought is on the the loading of the standard library from the install directory - I think we were considering moving dynamo's python install into

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 1]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
         /// Maps a regex to a particular python type.  Useful for matching things like dicts,
-        /// floats, strings, etc.  Initialized by 
+        /// floats, strings, etc.  Initialized by
         /// </summary>
         public Dictionary<string, Type> RegexToType = new Dictionary<string, Type>();
+        
+        /// <summary>
+        /// Maps a basic variable regex to a basic python type.
+        /// </summary>
+        public List<Tuple<Regex, Type>> BasicVariableTypes;
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: does this actually need to be public - can it be internal?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 2]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
+        
+        /// <summary>
+        /// Tracks already referenced CLR modules
+        /// </summary>
+        public HashSet<string> clrModules { get; set; }
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: internal?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 3]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
+        
+        /// <summary>
+        /// Keeps track of failed statements to avoid poluting the log
+        /// </summary>
+        public Dictionary<string, int> badStatements { get; set; }
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: internal?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 4]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
 
         /// <summary>
         /// A bunch of regexes for use in introspaction
         /// </summary>
-        public static string commaDelimitedVariableNamesRegex = @"(([0-9a-zA-Z_]+,?\s*)+)";
-        public static string variableName = @"([0-9a-zA-Z_]+(\.[a-zA-Z_0-9]+)*)";   
-        public static string doubleQuoteStringRegex = "(\"[^\"]*\")";
-        public static string singleQuoteStringRegex = "(\'[^\']*\')";
+        public static string commaDelimitedVariableNamesRegex = @"(([0-9a-zA-Z_]+,?\s?)+)";
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: same question for some of these members - I know that the previous ones were all public, but unless this is actually useful to someone making these public - makes refactoring the code between major releases very painful.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 5]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
+        public static string variableName = @"([0-9a-zA-Z_]+(\.[a-zA-Z_0-9]+)*)";
+        public static string quotesStringRegex = "[\"']([^\"']*)[\"']";
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: I think it would be very useful to add comments/summaries to some of these - especially if they are public to describe anything that is not obvious from the field names.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 6]
```diff
@@ -64,28 +60,60 @@ public ScriptScope Scope
...
+        public static string equalsRegex = @"(=)";
         public static string dictRegex = "({.*})";
         public static string doubleRegex = @"([-+]?[0-9]*\.?[0-9]+([eE][-+]?[0-9]+)?)";
         public static string intRegex = @"([-+]?\d+)[\s\n]*$";
         public static string basicImportRegex = @"(import)";
         public static string fromImportRegex = @"^(from)";
-
-#endregion
+        
+        private static readonly Regex MATCH_LAST_NAMESPACE = new Regex(@"[\w.]+$", RegexOptions.Compiled);
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: I don't have a problem with the readonly fields being uppercase - but curious - is this from a specific style guide?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 7]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
 
-                    scope.Engine.CreateScriptSourceFromString(revitImports, SourceCodeKind.Statements).Execute(scope);
+                    engine.CreateScriptSourceFromString(revitImports, SourceCodeKind.Statements).Execute(scope);
+                    clrModules.Add("RevitAPI");
+                    clrModules.Add("RevitAPIUI");
                 }
                 catch
                 {
-                    Log("Failed to load Revit types for autocomplete.  Python autocomplete will not see Autodesk namespace types.");
+                    Log("Failed to load Revit types for autocomplete. Python autocomplete will not see Autodesk namespace types.");
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: can we get more specific with this message?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 8]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
                 }
             }
 
-            if (assemblies.Any(x => x.FullName.Contains("ProtoGeometry")))
+            if (assemblies.Any(x => x.GetName().Name == "ProtoGeometry"))
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: why this change? speed?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 9]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
                 catch (Exception e)
                 {
                     Log(e.ToString());
-                    Log("Failed to load ProtoGeometry types for autocomplete.  Python autocomplete will not see Autodesk namespace types.");
+                    Log("Failed to load ProtoGeometry types for autocomplete. Python autocomplete will not see Autodesk namespace types.");
                 }
             }
-
+            
+            string pythonLibDir = System.IO.Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86),
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: I'm just thinking this might change soon -  but that should not hold up this pr.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 10]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
+                                                         "IronPython 2.7\\Lib");
+            if (System.IO.Directory.Exists(pythonLibDir) )
+            {
+                try
+                {
+                    var pyLibImports = String.Format("import sys\nsys.path.append(r'{0}')\n", pythonLibDir);
+                    engine.CreateScriptSourceFromString(pyLibImports, SourceCodeKind.Statements).Execute(scope);
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: I believe your previous work made the performance impact of this change minimal - is that true?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 11]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
-        /// Generates completion data for the specified text, while import the given types into the 
+        /// Generates completion data for the specified text, while import the given types into the
         /// scope and discovering variable assignments.
         /// </summary>
-        /// <param name="line">The code to parse</param>
+        /// <param name="code">The code to parse</param>
+        /// <param name="expand">Optionally recurse and extend the search to the entire namespace</param>
         /// <returns>Return a list of IronPythonCompletionData </returns>
-        public ICompletionData[] GetCompletionData(string line)
+        public ICompletionData[] GetCompletionData(string code, bool expand=false)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: we can't change this methods signature at this time - even adding an optional parameter will break binary compatibility:
https://stackoverflow.com/questions/1456785/a-definitive-guide-to-api-breaking-changes-in-net

It's very easy to miss cases here. :(

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 12]
```diff
@@ -97,116 +125,154 @@ public IronPythonCompletionProvider()
...
+            
+            else
+            {
+                Log("Valid IronPython installation not found. Python autocomplete will not see native modules.");
+            }
         }
 
         /// <summary>
-        /// Generates completion data for the specified text, while import the given types into the 
+        /// Generates completion data for the specified text, while import the given types into the
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: this comment does not seem grammatically correct.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 13]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
+                }
+            }
+            return statements;
         }
 
         /// <summary>
-        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.  
-        /// The ImportedTypes dictionary is 
+        /// Find all import statements and import into scope.  If the type is already in the scope, this will be skipped.
+        /// The ImportedTypes dictionary is
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: comment just ends :(

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 14]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
-                .Union(FindAllTypeImportStatements(code));
-
-            // try and load modules into python scope
-            foreach (var import in imports)
+            //detect all lib references prior to attempting to import anything
+            var refs = findClrReferences(code);
+            foreach (var statement in refs)
             {
-                if (scope.ContainsVariable(import.Key))
+                int previousTries = 0;
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: var

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 15]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
+                badStatements.TryGetValue(statement, out previousTries);
+                if (previousTries > 3)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: what is the significance of the 3? Can it be pulled out to a constant?

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 16]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
                 try
                 {
-                    scope.Engine.CreateScriptSourceFromString(import.Value, SourceCodeKind.SingleStatement)
-                         .Execute(this.scope);
-                    var type = Type.GetType(import.Key);
-                    this.ImportedTypes.Add(import.Key, type);
+                    string libName = MATCH_FIRST_QUOTED_NAME.Match(statement).Groups[1].Value;
+                    if (!clrModules.Contains(libName))
+                    {
+                        if (statement.Contains("AddReferenceToFileAndPath"))
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: please pull this string out as a constant

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 17]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
+                        {
+                            engine.CreateScriptSourceFromString(statement, SourceCodeKind.SingleStatement).Execute(scope);
+                            //it's an assembly path, don't check the current appdomain
+                            clrModules.Add(libName);
+                            continue;
+                        }
+                        
+                        if(AppDomain.CurrentDomain.GetAssemblies().Any(x => x.GetName().Name == libName))
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: hmm, maybe it is worth making a map of assemblies and checking if the name exists there instead of this loop - it appears lookup might happen a lot.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 18]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
                 {
-                    Console.WriteLine(exception.Message);
+                    Log(e.ToString());
+                    Log(String.Format("Failed to reference library: {0}", statement));
+                    badStatements[statement] = previousTries + 1;
+                }
+            }
+            
+            var importStatements = FindAllImportStatements(code);
+            foreach (var i in importStatements)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: can you add a comment about the general output of this for loop.

File: src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs [hunk 19]
```diff
@@ -597,131 +808,228 @@ protected Type TryGetType(string name)
...
+        
         /// <summary>
-        /// Returns the name from the end of a string.  Matches back to the first space and trims off spaces or ('s
-        /// from the end of the line
+        /// Returns the last name from the input line. The regex ignores tabs, spaces, the first new line, etc.
         /// </summary>
         /// <param name="text"></param>
         /// <returns></returns>
-        string GetName(string text)
+        string GetLastName(string text)
```
Comment (mjkkirschner) [inline src/Libraries/PythonNodeModelsWpf/IronPythonCompletionProvider.cs]: please add private access modifier

**Comment by Racel** [comment]
@dimven We think this looks pretty close and would like to get this into the next release. We would like to take this PR over, are you ok with that?

**Comment by dimven** [comment]
Hi @Racel & @mjkkirschner 

Sorry, I've been real bussy the last few weeks and haven't had a chance to look at this again. I would absolutely love it if you could take over and wrap it up for 2.1 :)

**Comment by alfarok** [comment]
Picking this work back up in #9402 as per the discussion above.  I am going to close this PR but feel free to continue any conversations/concerns in the new PR.
